Streamed fetch

MichalCz opened the pull request:
Here's a major rewrite of openaq-fetch.

### Major changes:

- simplification of the main `fetch.js` file by extracting all actual methods to library folder.
- separation of utility and fetch functions to separate files.
- introduction of streaming for database save and s3 streaming process
- more flexible adapters - measurements can be provided as any `streamable` object if exposing `fetchStream` method
- use of async/await in the fetch process
- compatibility with node 8
- additional documentation of formats in docs/
- additional command line options and switches

The `fetchStream` for an adapter is now much more flexible and can return (including asynchronously):

- `Array<Measurement>` - the same data that `fetchData` would return.
- `Iterable<Measurement>` - any iterable
- `AsyncGeneratorFunction<Measurement>` - in essence, `async function*(source) { ... }`, an [async generator](https://github.com/tc39/proposal-async-iteration#async-generator-functions)
- `Readable<Measutement>` - a readable stream that will be streamed to all endpoints.

### Benefits

The change introduces streaming throughout all the data fetching process from acquisition (on selected adapters) to database save and s3 streaming.

Here's the last output from a full fetch on an Intel i7 5820k, 16GB RAM machine:

        Command being timed: "node . -v"
        User time (seconds): 82.43
        System time (seconds): 4.96
        Percent of CPU this job got: 56%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 2:34.48
        Maximum resident set size (kbytes): 302696
        Minor (reclaiming a frame) page faults: 134670
        Voluntary context switches: 120780
        Involuntary context switches: 1538
        File system outputs: 8832
        Page size (bytes): 4096

Improvements include: 

- overall execution time - the measurements are saved to DB and streamed to s3 during fetch process.
- CPU usage - the transformations no longer operate on large arrays, there's also less garbage collection needed.
- memory usage - if the adapter fully supports streaming, the data is never aggregated and kept in memory as whole.

jflasher commented:
Thanks @MichalCz, I'll be checking this out over the next few days!

MichalCz commented:
Ok. In mean time I did find some issues with Error handling, that is causing the program to exit despite not having set the `strict` flag nor env...

jflasher commented on lib/measurement.js:
```diff
@@ -0,0 +1,227 @@
+  return createFetchObject(out, source, dryrun);
+}
+
+export function printDataToLog (stream, log) {
```
I noticed when running `node index.js -d -s 'Beijing US Embassy' -v` I'm seeing output like 

```
2018-09-10T12:45:18.742Z - verbose:  utc=Mon Sep 10 2018 08:00:00 GMT-0400 (EDT), local=2018-09-10T20:00:00+08:00, parameter=pm25, value=41, unit=µg/m³, value=1, unit=hours, location=Beijing US Embassy, city=Beijing, country=CN, latitude=39.95, longitude=116.47, attribution=[name=StateAir.net, url=http://www.stateair.net/web/post/1/1.html], sourceName=Beijing US Embassy, sourceType=government, mobile=false
```
which is not stringified. For this output, I prefer stringified as it'll convert the UTC time properly which makes it easier to examine (instead of it taking into account your local time zone when it's an object). The output at this code block is being stringified, but this must not be where this output is happening for the verbose logging.

MichalCz replied:
Hmm... the "why" is quite simple - the schema says that one should be a Date object, second a String:

```
        "date": {
            "type": "object",
            "additionalProperties": false,
            "required": ["utc", "local"],
            "properties": {
                "utc": {
                    "type": "date"
                },
                "local": {
                    "type": "string",
                    "pattern": "\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}(\\+|-)\\d{2}:\\d{2}"
                }
            }
        },
```

So when that's stringified you get different valuies.

I'll check if that doesn't pose any additional issues, but could we simply use two strings here? The database should be ok with that and the strings will be the same - and I'll add mapping for old format.

MichalCz replied:
We would, but I simply added a date correction step along the way - so no need to change any adapter code.

In general I'm trying to make no changes at all to __all adapters__. Whatever worked before should work after.

MichalCz commented:
Noted, working on all of those. :)

MichalCz commented:
Ok, so one thing to add to the list above - there's an issue with how Babel works out if an object is `instanceof` something derived from `Error` - the issue is that it doesn't work at all.

So now sadly any Error is simply thrown and I need to make a small workaround. After that I'll get to the list above.

MichalCz commented on docs/source.md:
```diff
@@ -0,0 +1,39 @@
+* `name (String)` - an *informative* source name
+* `country (String)` - the country for which the data is provided
+* `description (String)` - description of the source
+* `contacts (Array<String>)` - a list of e-mail addresses for contact about the source
```
I can set a default value of that if you like, but if you'd want to use that field in future - I personally would leave it as is.

jflasher replied:
Yeah, I think it's fine as is. When I originally read this, the description made me think it was contacts at the source of the data. But it looks fine to me.

MichalCz commented on docs/source.md:
```diff
@@ -0,0 +1,39 @@
+
+Optional properties (used to add information to measurements):
+
+* `type (Enum<government|private|?>)` - Region name
```
Thanks. Will fix this. :)

MichalCz commented on docs/source.md:
```diff
@@ -0,0 +1,39 @@
+
+Additonally the whole object is passed to the adapter so anything can be added to the structure. Adapter usually use:
+
+* `sourceURL (String)` - the url to use to fetch the data from
```
Ok. Will fix.

MichalCz commented on fetch.js:
```diff
@@ -7,451 +7,80 @@
+      // mark sources as finished
+      .do(markSourceAs('finished', runningSources))
+      // handle adapter errors to be forwarded to main stream and well handled.
+      .use(forwardErrors)
```
First of all - the method here is supposed to selectively forward errors from the measurements stream flowing from adapter to S3, database and console, to the stream that runs all adapters. Any error forwarded there gets handled by the `handleFetchErrors` method at the end of the file.

I'll document this a bit, but in general all errors coming from an adapter are handled as stream errors, that is:
 * an adapter fetch error (thrown as the FetchError class).
 * an asynchronous error.
 * wrong data format returned (not an array, data stream, generator etc.)
 * adapter not found.
 * measurement error.

In effect it should:
 - ignore adapter fetch errors (so simply mute them here and, after a small change I need to commit still, also log a warning)
 - ignore measurement errors and only log them.

As previously it won't ignore:
 - adapter runtime errors
 - adapter syntax errors
these are handled, but forwarded to the main stream.

This only changes if you turn on `--strict` mode, where any error should be forwarded to the main stream.

MichalCz commented on fetch.js:
```diff
@@ -7,451 +7,80 @@
+    handleFetchErrors(log, env)
+  )
+  .then(
+    exitCode => process.exit(exitCode || 0)
```
Actually no, `handleFetchErrors` extracts the exitCode from the `FetchError` object - if there's something there it will be that (for instance `100`), but if that's not given a default of `0` is used.

MichalCz commented on fetch.js:
```diff
@@ -7,451 +7,80 @@
+      .toArray()
+      // save fetch log to DB and send a webhook if necessary.
+      .then(
+        // TODO: filter out sources not included in fetch
```
Yes, the ones that were made `filtered` - I forgot to ask you about this.

So when we run a single source - what should we save to the db?
And when we run only active sources - should we save only those?

So I guess we should remove them, but I had some second thoughts back then.

jflasher replied:
I think whatever gets run should get saved. I only ever run a single source for testing purposes, but would want just that source saved to the db. And if I'm only running active sources, would only want those saved.

MichalCz replied:
Added.

MichalCz commented on lib/utils.js:
```diff
@@ -156,3 +68,65 @@ export function safeParse (json) {
+    const cause = error instanceof FetchError ? error : error.cause;
+
+    if (cause instanceof FetchError) {
+      if (cause.is(STREAM_END)) return cause.exitCode || 0;
```
I hear you. :)

I think I actually don't normally send stacks to `Winston` now, but that's worth double checking. Only place where that happens is in the final error handler, before process.exit.

In any case I'll double check and report back.

MichalCz replied:
No stacks are going to Winston except for when fetch.js actually crashes.

MichalCz replied:
And that's a single stack, so it should be ok to leave it.

MichalCz commented on lib/utils.js:
```diff
@@ -1,134 +1,46 @@
-      continue;
-    }
+export function convertUnits (array) {
+  array.filter(x => x).forEach(unifyMeasurementUnits);
```
`x => x` in `filter` is simply to filter out falsy values, especially nulls or undefined. Those units cannot be unified, since there's nothing there to unify. ;)

BTW, I've taken the liberty of removing this method in a next commit (I left a stub, so that legacy adapters won't break). Instead it's being run by the fetch process, along with a couple other checks (like removing unwanted measurements).

MichalCz commented on lib/env.js:
```diff
@@ -0,0 +1,112 @@
+  .usage('Usage: $0 -d -s \'Beijing US Embassy\'')
+  .options('quiet', {
+    boolean: true,
+    describe: 'Show lots additional logging information (more than verbose)',
```
Very much so... ;)

MichalCz commented:
Joe, I need to push the dependencies update in the same pull request otherwise we'd start a major conflict and that would get worse and worse. So here's another group of commits.

MichalCz commented on fetch.js:
```diff
@@ -7,451 +7,80 @@
  */
```
Hmm... Strangely I can't see 2048, but 2368 in my code. Let me push a couple new commits and let's see...

jflasher replied:
I'm now seeing 2688 for new code, 2899 for old code. Wonder if this is due to stricter rejection code somewhere?

MichalCz replied:
That is possible - I did fiddle with the quality checks quite a lot, so it may be I missed some kind of lowercase. I'll check that.

MichalCz replied:
It seems there has been some issues with `.batch(64)` in the adapter itself - it actually is a regression in `scramjet`.

I removed the flatmap as it's not needed anymore so I'll have more time looking for the batch issue.

MichalCz commented on .build_scripts/bench-adapters.js:
```diff
@@ -0,0 +1,9 @@
+'use strict';
```
Actually I added that in the .es6 module...

MichalCz commented:
I need to rewrite the measurement test a little so it works again. I'll finish that tomorrow. In meanwhile if you could check my answers that'd be good. :)

MichalCz commented:
Joe - I've added ppb as measurement unit - was that intentionally left out as a difference between long billion (thousand millards) and short billion (thousand millions)?

jflasher commented:
> @jflasher - I've added ppb as measurement unit - was that intentionally left out as a difference between long billion (thousand millards) and short billion (thousand millions)?

How do you mean you added it? At the end of the day, we should only end up with measurements using units of either `µg/m³` or `ppm`. If something comes in like `ug/m3`, `PPM`, we can do a simple name change. If it's like `g/cm3` or `ppb` or variants, we will do some math to convert to our preferred units. We don't convert across `µg/m³` and `ppm` and if something has a unit we don't understand, we just ignore it (or at least we should). The renaming and math were done in https://github.com/openaq/openaq-fetch/blob/develop/lib/utils.js#L92

MichalCz commented:
@jflasher I mean I added the ppb to ppm conversion. But if I read you correctly I will revert that change, please confirm.

jflasher commented:
I think this is probably fine as long as it’s done at some point before the final validation occurrs so we don’t lose any measurements because of it. And if you could keep the same conversion math as was in the previous method, that’d be great. While the math isn’t hard, it care out of looking at a bunch of the ways different sites did their labeling which is often a mess!


On September 13, 2018 at 18:21:47, Budleigh Salterton (notifications@github.com(mailto:notifications@github.com)) wrote:

>  
> Actually to make it clearer: all adapters used a function called convertUnits. I made this function (or it's successor) part of the main flow which now automatically coverts pphm to ppm and mg to µg.
>  
>  
> At the same time the convertion by convertUnits was removed as not needed and done later in the flow.
>  
>  
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub(https://github.com/openaq/openaq-fetch/pull/520#issuecomment-421171496), or mute the thread(https://github.com/notifications/unsubscribe-auth/AAz0JudI5ScgZHvD67k5v7fkUuOAikRrks5uatp7gaJpZM4WbBnq).
>  
>  
>

MichalCz commented:
@jflasher done. :)

MichalCz commented:
@jflasher please hold off the pull for another day or two. I've noticed as problem in error handling in scramjet and I'd need to fix it. Currently the errors are either completely muted or crash the fetch process.

Because of a small mistake, scramjet doesn't pass the error wrapped in a transport object when it occurs in another error handler. Instead it simply passes the actual error which is inconsistent, I did however rely on this to work. I'll release the new version of scramjet tomorrow and update this branch.

MichalCz commented:
Hello @jflasher,

All the fixes are now in. I've added some docs and removed some unused code left over from the refactor.

The adapters now nicely report errors. This is also now covered by automated tests.

jflasher commented on lib/db.js:
```diff
@@ -0,0 +1,173 @@
+ * @param {object} m measurement object
+ * @return {object} an object capable of being saved into the PostgreSQL database
+ */
+function convertMeasurementToSQLObject (stream, pg) {
```
Hi @MichalCz, looking over this this morning, and saw the following behavior when trying to run using the local database.

```
$ node index.js -s 'StateAir_Hanoi'                                                               [±:signicode-feature/streamed-libs:∂]
2018-09-24T11:58:49.974Z - info: --- Full fetch started. ---
2018-09-24T11:58:49.988Z - info: Looking up source StateAir_Hanoi
2018-09-24T11:58:50.547Z - info: Getting stream for "StateAir_Hanoi" from "stateair"
2018-09-24T11:58:51.018Z - info: Using connection data from knexfile.local
2018-09-24T11:58:51.112Z - info: Connecting to database.
2018-09-24T11:58:51.338Z - info: Database connected and migrations are handled, ready to roll!
2018-09-24T11:58:51.374Z - error:  message=insert into "measurements" ("attribution", "averagingPeriod", "city", "coordinates", "country", "date", "location", "mobile", "parameter", "sourceName", "sourceType", "unit", "value") values ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13) returning "location" - column "attribution" of relation "measurements" does not exist, stack=error: column "attribution" of relation "measurements" does not exist
```

This is odd because https://github.com/openaq/openaq-fetch/pull/520/files#diff-be228df81b99a384aa50ad50bde7fca0R38 looks to be doing what I expect it should be doing, but the measurement trying to be inserted doesn't seem to be hitting it? Note how it doesn't include `date_utc` or `data` but does include `attribution`.

MichalCz replied:
I'll check that... this is the first column so it may be that the whole table is somehow badly created... Maybe my knexfile-local.json causes that? Or db is not right? I'll get back to you on this.

MichalCz replied:
Fixed.

jflasher commented on docs/dev-quick-start.md:
````diff
@@ -0,0 +1,35 @@
+CREATE EXTENSION postgis;
+\quit
+```
+6. Create a local knexfile `cp knexfile.js knexfile.local.js && editor knexfile.local.js`
````
Change to `knexfile-local.js`

MichalCz replied:
Doing

MichalCz replied:
Done.

jflasher commented on docs/index.md:
```diff
@@ -0,0 +1,15 @@
+# Documentation
+
```
Can you add a link to this somewhere in the README file? This is incredibly helpful and I don't want people to miss it.

MichalCz replied:
Doing.

MichalCz replied:
Done.

jflasher commented:
I was using my existing db and the migrations didn’t change so I’m thinking it has something to do with the way the object to be inserted is created rather than anything on the db side?


On September 24, 2018 at 09:06:01, Budleigh Salterton (notifications@github.com(mailto:notifications@github.com)) wrote:

>  
> @MichalCz commented on this pull request.
>  
>  
> In lib/db.js(https://github.com/openaq/openaq-fetch/pull/520#discussion_r219827412):
>  
> > + })() + .then(async pg => { + await pg.migrate.latest(knexConfig); + log.info('Database connected and migrations are handled, ready to roll!'); + return pg; + }); + // Run any needed migrations and away we go + return _pg; +} + +/** + * Build an object that can be inserted into our database. + * @param {object} m measurement object + * @return {object} an object capable of being saved into the PostgreSQL database + */ +function convertMeasurementToSQLObject (stream, pg) {  
>  
> I'll check that... this is the first column so it may be that the whole table is somehow badly created... Maybe my knexfile-local.json causes that? Or db is not right? I'll get back to you on this.
>  
>  
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub(https://github.com/openaq/openaq-fetch/pull/520#discussion_r219827412), or mute the thread(https://github.com/notifications/unsubscribe-auth/AAz0JnlN6IjJ_IEM9Plw_zdkgUL-DAZJks5ueNi5gaJpZM4WbBnq).
>  
>  
>

MichalCz commented:
You may be right. I recall there was a wrapping function somewhere there. I'll check it.

MichalCz commented:
@jflasher I also added a small change to result messages pushed out to logs (including number of duplicates and time elapsed).

jflasher commented on lib/db.js:
```diff
@@ -0,0 +1,182 @@
+ * @param {object} m measurement object
+ * @return {object} an object capable of being saved into the PostgreSQL database
+ */
+function convertMeasurementToSQLObject (stream, pg) {
```
Trying the latest, when running with command below, I'm getting an error.

```
$ node index.js -s 'StateAir_Hanoi' -d                                                            [±:signicode-feature/streamed-libs:∂]
2018-09-26T13:41:29.746Z - info: --- Dry run for Testing, nothing is saved to the database. ---
2018-09-26T13:41:29.775Z - info: Looking up source StateAir_Hanoi
2018-09-26T13:41:30.276Z - info: Getting stream for "StateAir_Hanoi" from "stateair"
2018-09-26T13:41:30.685Z - info: Using connection data from knexfile-local
2018-09-26T13:41:30.756Z - info: Connecting to database.
2018-09-26T13:41:30.958Z - info: Database connected and migrations are handled, ready to roll!
2018-09-26T13:41:31.164Z - error: Runtime error occurred in DataStream(unused): Error: Cannot read property 'utc' of undefined
    at Promise.all.catch (/Users/jflasher/workspace/openaq-fetch/node_modules/scramjet-core/lib/util/mk-transform.js:59:44)
    at <anonymous>
  caused by:
    at DataStream.stream.map.data (/Users/jflasher/workspace/openaq-fetch/lib/db.js:50:17)
    at <anonymous>
  --- raised in DataStream(unused) constructed ---
    at PromiseTransformStream (/Users/jflasher/workspace/openaq-fetch/node_modules/scramjet-core/lib/util/promise-transform-stream.js:65:27)
    at DataStream (/Users/jflasher/workspace/openaq-fetch/node_modules/scramjet-core/lib/data-stream.js:55:9)
    at DataStream.map (/Users/jflasher/workspace/openaq-fetch/node_modules/scramjet-core/lib/data-stream.js:153:26)
    at convertMeasurementToSQLObject (/Users/jflasher/workspace/openaq-fetch/lib/db.js:41:17)
    at DataStream.use (/Users/jflasher/workspace/openaq-fetch/node_modules/scramjet-core/lib/data-stream.js:304:20)
    at streamRecordsToPg (/Users/jflasher/workspace/openaq-fetch/lib/db.js:75:6)
    at DataStream.use (/Users/jflasher/workspace/openaq-fetch/node_modules/scramjet-core/lib/data-stream.js:304:20)
    at streamDataToDB (/Users/jflasher/workspace/openaq-fetch/lib/db.js:93:6)
    at /Users/jflasher/workspace/openaq-fetch/lib/db.js:178:9
    at <anonymous>
```

I put a log statement to see what the `data` variable was getting. In the case where it errors, it's getting the below which looks like the already converted object. So it looks like this function may be getting called multiple times and will fail if it runs on already converted data, which I suppose is good! 

Note also that the coordinates on the converted object is the full postgis coordinates object, but just needs to be a simple lat/lon object.

```
{ location: 'US Diplomatic Post: Hanoi',
  value: 17,
  unit: 'µg/m³',
  parameter: 'pm25',
  country: 'VN',
  city: 'Hanoi',
  source_name: 'StateAir_Hanoi',
  date_utc: '2018-09-26T12:00:00.000Z',
  source_type: 'government',
  mobile: false,
  data: 
   { date: 
      { utc: '2018-09-26T12:00:00.000Z',
        local: '2018-09-26T19:00:00+07:00' },
     parameter: 'pm25',
     value: 17,
     unit: 'µg/m³',
     averagingPeriod: { value: 1, unit: 'hours' },
     location: 'US Diplomatic Post: Hanoi',
     city: 'Hanoi',
     country: 'VN',
     coordinates: { latitude: 21.021938, longitude: 105.81881 },
     attribution: [ [Object] ],
     sourceName: 'StateAir_Hanoi',
     sourceType: 'government',
     mobile: false },
  coordinates: 
   Raw {
     client: 
      Client_PG {
        config: [Object],
        logger: [Object],
        connectionSettings: [Object],
        driver: [Object],
        pool: [Object],
        valueForUndefined: [Object],
        _events: [Object],
        _eventsCount: 4,
        makeKnex: [Function: makeKnex],
        version: '9.5.3' },
     sql: 'ST_geomFromText(?, ?)',
     bindings: [ 'Point(105.81881 21.021938)', 4326 ],
     _wrappedBefore: undefined,
     _wrappedAfter: undefined,
     _debug: undefined } }
```

MichalCz replied:
I'll check it today or tomorrow evening (should be fixed tomorrow your time).

MichalCz replied:
Oh, now I see my misunderstaning. It's the `-d` - a small fix resolved this.

MichalCz replied:
So this was due to the fact that I removed an unneeded argument from one of the methods and didn't change the call. Now this will work as it should.

jflasher replied:
Just dropped the `-d` and it's still happening, same error.

MichalCz replied:
I found it now. It didn't occur if the data was not streamed to S3. As you suspected the object was transformed to the SQL version twice.

jflasher replied:
Hi @MichalCz, I wanted to make sure I should be waiting for another commit before checking fetch behavior? I didn't see any commit since your last comment and wanted to make sure I didn't miss something. Thanks!

MichalCz replied:
Hi @jflasher I didn't push to the right remote after I fixed the issue. :(

There's that commit now.

jflasher replied:
Great, this looks to be working now, thanks!

jflasher commented:
Thanks!


On September 26, 2018 at 07:29:49, Budleigh Salterton (notifications@github.com(mailto:notifications@github.com)) wrote:

>  
> @MichalCz commented on this pull request.
>  
>  
> In lib/db.js(https://github.com/openaq/openaq-fetch/pull/520#discussion_r220586650):
>  
> > + })() + .then(async pg => { + await pg.migrate.latest(knexConfig); + log.info('Database connected and migrations are handled, ready to roll!'); + return pg; + }); + // Run any needed migrations and away we go + return _pg; +} + +/** + * Build an object that can be inserted into our database. + * @param {object} m measurement object + * @return {object} an object capable of being saved into the PostgreSQL database + */ +function convertMeasurementToSQLObject (stream, pg) {  
>  
> I'll check it today or tomorrow evening (should be fixed tomorrow your time).
>  
>  
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub(https://github.com/openaq/openaq-fetch/pull/520#discussion_r220586650), or mute the thread(https://github.com/notifications/unsubscribe-auth/AAz0JutvlphLcDdJfN7sEDbIoJ86vR7mks5ue49cgaJpZM4WbBnq).
>  
>  
>

MichalCz commented:
@jflasher I added test for the following command: `node . -ds test_adapter` to the tests run. This way a mistake like the one fixed yesterday should not go past the tests. This way there will also be a simple confirmation if the fetch process still works correctly after any new changes.

jflasher commented on lib/db.js:
```diff
@@ -0,0 +1,181 @@
+      const pg = await getDB();
+
+      await Promise.all([
+        doSaveToS3 ? saveResultsToS3(stream, new (require('aws-sdk').S3)(), bucketName, key) : true,
```
Was looking to test the data format of data saved to S3. Ran with the following command (using valid AWS key/secret)

`AWS_ACCESS_KEY_ID=ID AWS_SECRET_ACCESS_KEY=KEY AWS_BUCKET_NAME=openaq-fetches SAVE_TO_S3=true node index.js` 

and was seeing a lot of the following errors

```
(node:29353) UnhandledPromiseRejectionWarning: Error: Cannot read property 'Symbol(Symbol.iterator)' of undefined
    at promise.catch (/Users/jflasher/workspace/openaq-fetch/node_modules/scramjet-core/lib/util/promise-transform-stream.js:221:31)
    at process._tickCallback (internal/process/next_tick.js:178:7)
  caused by:
    at into (/Users/jflasher/workspace/openaq-fetch/node_modules/scramjet/lib/data-stream.js:380:35)
    at process._tickCallback (internal/process/next_tick.js:178:7)
  --- raised in DataStream(unused) constructed ---
    at new PromiseTransformStream (/Users/jflasher/workspace/openaq-fetch/node_modules/scramjet-core/lib/util/promise-transform-stream.js:65:27)
    at new DataStream (/Users/jflasher/workspace/openaq-fetch/node_modules/scramjet-core/lib/data-stream.js:55:9)
    at DataStream.flatMap (/Users/jflasher/workspace/openaq-fetch/node_modules/scramjet/lib/data-stream.js:373:21)
    at saveResultsToS3 (/Users/jflasher/workspace/openaq-fetch/lib/db.js:108:6)
    at /Users/jflasher/workspace/openaq-fetch/lib/db.js:176:22
    at process._tickCallback (internal/process/next_tick.js:178:7)
(node:29353) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 186533)
```

Tried again without the extra `SAVE_TO_DB` and `AWS_BUCKET_NAME` vars and everything worked fine again, so I presume it's tied to the output there. What's interesting is that this didn't happen at the end of the fetch which is when I presume the data is getting written to S3, but happened in mid-fetch as the adapters were running.

MichalCz replied:
The happening in middle part is expected - steaming to s3 happens white adapters were still running. I'll check this out tomorrow.

MichalCz replied:
Ok. For some reason I didn't merge some of the work from the s3 branch and that was an old version which didn't have a slight chance to work.

@jflasher - there's a limitation of s3 that requires the uploads to have file sizes set before the upload - trouble is, we don't actually know how big the file will be. Therefore the stream is chunked. In the upload code I set a chunk to 1MB - which sets a maximum of 10GB per file (S3 allows a maximum of 10k chunks).

Now I thought this should be controller via env, so let me add a quick commit for that.

MichalCz replied:
Ok, that's also pushed.

MichalCz replied:
Yeah. I did note that. So I'm rewriting the save to s3 so that all the measurements are sent to a single stream constructed before the fetch is started - and that stream will get saved.

I got this working, but at the same time I'll need to spend little more time to make it look better.

MichalCz commented on lib/db.js:
```diff
@@ -0,0 +1,181 @@
+      time_started: new Date(timeStarted),
+      time_ended: new Date(timeEnded),
+      count: itemsInserted,
+      results: JSON.stringify(errors || results)
```
I guess I wanted to get too much info there. Will look into this tomorrow.

MichalCz replied:
Hi Joe,

Indeed the alternative `(errors || results)` was out of place (especially that the errors are included the results object). This has been fixed, postgre now shows:

```sql
select * from fetches;

-[ RECORD 1 ]+--------------------------------------------------------------------------------------------------------------------------------------
id           | 49
time_started | 2018-10-04 14:14:48.997+00
time_ended   | 2018-10-04 14:14:51.324+00
count        | 24
results      | [{"message":"New measurements found for StateAir_Hanoi: 24","failures":{},"count":24,"duration":0.266,"sourceName":"StateAir_Hanoi"}]
```

jflasher commented:
Ah right, forgot that we were streaming here!


On October 3, 2018 at 19:38:32, Budleigh Salterton (notifications@github.com(mailto:notifications@github.com)) wrote:

>  
> @MichalCz commented on this pull request.
>  
>  
> In lib/db.js(https://github.com/openaq/openaq-fetch/pull/520#discussion_r222500099):
>  
> > + * + * @param {OpenAQEnv} env + * @param {String} bucketName + */ +export function streamMeasurementsToDBAndStorage ({doSaveToS3, dryrun}, bucketName) { + return async ({ stream, counts }) => { + if (dryrun) { + return stream + .do(m => log.verbose(JSON.stringify(m))) + .run(); + } else { + const key = `realtime/${moment().format('YYYY-MM-DD/X')}.ndjson`; + const pg = await getDB(); + + await Promise.all([ + doSaveToS3 ? saveResultsToS3(stream, new (require('aws-sdk').S3)(), bucketName, key) : true,  
>  
> The happening in middle part is expected - steaming to s3 happens white adapters were still running. I'll check this out tomorrow.
>  
>  
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub(https://github.com/openaq/openaq-fetch/pull/520#discussion_r222500099), or mute the thread(https://github.com/notifications/unsubscribe-auth/AAz0JlyG0CCnSO60eCG4_Y_wBetzYLqLks5uhUp4gaJpZM4WbBnq).
>  
>  
>

MichalCz commented on lib/db.js:
```diff
@@ -162,8 +177,10 @@ export async function saveSources (sources) {
  */
-export function streamMeasurementsToDBAndStorage ({doSaveToS3, dryrun}, bucketName) {
-  return async ({ stream, counts }) => {
+export function streamMeasurementsToDBAndStorage (stream, {doSaveToS3, doSaveToDB, dryrun}, bucketName) {
```
Now all the data is saved indeed. I'll chain s3 save after db save and add a filter for duplicates. ETA today.

MichalCz replied:
Fixed in the last commit. Only saved items are being pushed to s3.

MichalCz commented on lib/env.js:
```diff
@@ -96,6 +96,7 @@ export const getEnv = () => {
   const processTimeout = _env.PROCESS_TIMEOUT || 10 * 60 * 1000; // Kill the process after a certain time in case it hangs
   const bucketName = _env.AWS_BUCKET_NAME || '';
   const doSaveToS3 = _env.SAVE_TO_S3 === 'true' || +_env.SAVE_TO_S3;
+  const doSaveToDB = !('SAVE_TO_DB' in _env) || _env.SAVE_TO_DB === 'true' || +_env.SAVE_TO_DB;
```
I used that for testing only. But your other comment clears this out. I'll remove this.

MichalCz replied:
Removed.

MichalCz commented:
@jflasher as promised here's the updated version of s3 save with:

* Only one s3 file per fetch with all data
* S3 file is now closed at the end, which required me to add a new, rather helpful thing: `errors.js:cleanup.add` - calling this method will allow you to run any asynchronous function before the process exits (for example: make sure that s3 is saved before we call process.exit()). May come in handy in other cases.
* Added small improvements in log messages in debug mode.
* Removed `doSaveToDB` toggle

jflasher commented on lib/db.js:
```diff
@@ -110,11 +111,25 @@ function streamDataToDB (stream, pg, counts) {
     .catch(e => output.raise(e));
 
-  output.whenEnd().then(() => stream.end()).catch(ignore);
+  const endStream = () => {
```
I'm still seeing a timeout when running the following command

`AWS_ACCESS_KEY_ID=ID AWS_SECRET_ACCESS_KEY=KEY AWS_BUCKET_NAME=openaq-fetches SAVE_TO_S3=true node index.js`

there are no running adapters and looking at the db, there are no new measurements being inserted for a long time before timeout, so I wonder if this is still not getting called?

MichalCz replied:
Not this actually, but I have a hunch...

MichalCz replied:
Hi Joe - I found the issue in an implementation of `empty` method in scramjet. I should fix it by tomorrow morning your time.

MichalCz replied:
Hi again. Just to let you know, the issue I found seems more complex to solve because of conflicting requirements in couple different methods - so I'm working on this currently, but it's taking more than than anticipated.

MichalCz replied:
Hi Joe, the commit I just pushed should fix the issue.

MichalCz commented:
Hi @jflasher,

How's the check going? Did you manage to run this version on staging?

Best,
M.

jflasher commented:
Sorry @MichalCz was disconnected for the last few days, will look at this in the next couple of days.

jflasher commented:
Saving to database and S3 looks fine, so I'm going to merge this into staging and see how it runs there.
